modeling/algorithm/local-weighted-slice-algorithm.go

```diff
@@ -0,0 +1,211 @@
+// 'borrows' and 'rents' endpoints from other zones to make the local
+// EndpointSlice balanced with the incoming traffic. This variation uses weights
+// to make precise distribution without float to int approximation
+type LocalWeightedSliceAlgorithm struct{}
+
+// CreateSliceGroups creates sliceGroups with weights to indicate float
+// endpoints. Zones will have local sliceGroup representing integer number of
+// endpoints while having a shared sliceGroup with weights representing float
+// number of endpoints
+func (alg LocalWeightedSliceAlgorithm) CreateSliceGroups(region types.RegionInfo) (map[string]types.EndpointSliceGroup, error) {
+	if region.ZoneDetails == nil {
+		return nil, fmt.Errorf("zoneDetail should not be nil")
+	}
+	sliceGroups := map[string]types.EndpointSliceGroup{}
+	// endpointsAvailable stores zones with int number of endpoints available
+	endpointsAvailable := endpointsList{}
+	// endpointsNeeded stores zones with int number of endpoints needed
+	endpointsNeeded := endpointsList{}
+	// weightedEndpointsAvailable stores zones with float number of endpoints
+	// available
+	weightedEndpointsAvailable := endpointsList{}
+	// weightedEndpointsNeeded stores zones with float number of endpoints
+	// needed
+	weightedEndpointsNeeded := endpointsList{}
+	for zoneName, zone := range region.ZoneDetails {
+		var localGroup types.EndpointSliceGroup
+		localGroup.Label = zoneName
+		// this local sliceGroup should only receive traffic from current zone,
+		// this map tracks weights of traffic from different zones to this
+		// sliceGroup
+		localGroup.ZoneTrafficWeights = map[string]float64{zoneName: 1.0}
+		// this map stores composition of this sliceGroup, zoneName - number of
+		// endpoints from zoneName
+		localGroup.Composition = map[string]types.WeightedEndpoints{}
+
+		// calculate expected number of endpoints based on the proportion of
+		// nodes in this zone
+		expectedEndpoints := zone.NodesRatio * float64(region.TotalEndpoints)
+		// deviation: a negative value means this zone needs more endpoints from
+		// other zones, a positive value means this zone needs to give out
+		// endpoints to other zones
+		deviation := float64(zone.Endpoints) - expectedEndpoints
+		// intDeviation is dealt same way as original local-slice algorithm that
+		// directly 'borrows' or 'sends' endpoints among zones
+		intDeviation := int(deviation)
+		if intDeviation == 0 {
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: int(expectedEndpoints), Weight: 1}
+		} else if intDeviation > 0 {
+			endpointsAvailable.push(endpointDeviation{name: zoneName, deviation: intDeviation})
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: int(expectedEndpoints), Weight: 1}
+		} else {
+			endpointsNeeded.push(endpointDeviation{name: zoneName, deviation: -intDeviation})
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: zone.Endpoints, Weight: 1}
```

Comment:
I had a similar comment above that these WeightedEndpoints could likely be declared above the if statements to reduce the duplication.

Reply:
changed